examples/jsm/renderers/webgpu/WebGPURenderer.js

```diff
@@ -0,0 +1,601 @@
+
+							}
+
+						}
+
+					} else if ( material.visible ) {
+
+						currentRenderList.push( object, geometry, material, groupOrder, _vector3.z, null );
+
+					}
+
+				}
+
+			}
+
+		}
+
+		const children = object.children;
+
+		for ( let i = 0, l = children.length; i < l; i ++ ) {
+
+			this._projectObject( children[ i ], camera, groupOrder );
+
+		}
+
+	}
+
+	_renderObjects( renderList, camera ) {
+
+		for ( let i = 0, l = renderList.length; i < l; i ++ ) {
+
+			const renderItem = renderList[ i ];
+
+			const object = renderItem.object;
+
+			object.modelViewMatrix.multiplyMatrices( camera.matrixWorldInverse, object.matrixWorld );
+			object.normalMatrix.getNormalMatrix( object.modelViewMatrix );
+
+			this._objects.update( object );
+			this._bindings.update( object, camera );
+
+			this._renderObject( object );
+
+		}
+
+	}
+
+	_renderObject( object ) {
+
+		const device = this._device;
+		const info = this._info;
+
+		// begin
+
+		const cmdEncoder = device.createCommandEncoder( {} );
+		const passEncoder = cmdEncoder.beginRenderPass( this._renderPassDescriptor );
+
+		// pipeline
+
+		const pipeline = this._renderPipelines.get( object );
+		passEncoder.setPipeline( pipeline );
+
+		// rasterization
+
+		const vp = this._viewport;
+
+		if ( vp !== null ) {
+
+			const width = Math.floor( vp.width * this._pixelRatio );
+			const height = Math.floor( vp.height * this._pixelRatio );
+
+			passEncoder.setViewport( vp.x, vp.y, width, height, vp.minDepth, vp.maxDepth );
+
+		}
+
+		const sc = this._scissor;
+
+		if ( sc !== null ) {
+
+			const width = Math.floor( sc.width * this._pixelRatio );
+			const height = Math.floor( sc.height * this._pixelRatio );
+
+			passEncoder.setScissorRect( sc.x, sc.y, width, height );
+
+		}
+
+		// bind group
+
+		const bindGroup = this._bindings.get( object ).group;
+		passEncoder.setBindGroup( 0, bindGroup );
+
+		// index
+
+		const geometry = object.geometry;
+		const index = geometry.index;
+
+		const hasIndex = ( index !== null );
+
+		if ( hasIndex === true ) {
+
+			this._setupIndexBuffer( passEncoder, index );
+
+		}
+
+		// vertex buffers
+
+		this._setupVertexBuffers( passEncoder, geometry.attributes );
+
+		// draw
+
+		const drawRange = geometry.drawRange;
+		const firstVertex = drawRange.start;
+
+		if ( hasIndex === true ) {
+
+			const indexCount = ( drawRange.count !== Infinity ) ? drawRange.count : index.count;
+
+			passEncoder.drawIndexed( indexCount, 1, firstVertex, 0, 0 );
+
+			info.update( object, indexCount );
+
+		} else {
+
+			const positionAttribute = geometry.attributes.position;
+			const vertexCount = ( drawRange.count !== Infinity ) ? drawRange.count : positionAttribute.count;
+
+			passEncoder.draw( vertexCount, 1, firstVertex, 0 );
+
+			info.update( object, vertexCount );
+
+		}
+
+		// end
+
+		passEncoder.endPass();
+		device.defaultQueue.submit( [ cmdEncoder.finish() ] );
```

Comment:
Comment for future optimization. As you may be already aware, in my experience I found that creating command encoder, calling `.beginRenderPass` and `.setPipeline`, and submitting default queue for each object seems costly (in current WebGPU implementation). 

I tried to avoid unnecessary creation and call, the performance became much better.

https://github.com/takahirox/THREE.WebGPURenderer/blob/master/src/renderers/WebGPURenderer.js#L568-L624
https://github.com/takahirox/THREE.WebGPURenderer/blob/master/src/renderers/WebGPURenderer.js#L504-L507

Reply:
Yes, good point. I also wanted to experiment with bundles and see how it affects performance.

Reply:
Should be more efficient now: 9021c89b60e3505924d22e2340c3a05a121ba2b4

A call of `WebGPURenderer.render()` can be considered as a single render pass. This notion maps nicely on the WebGPU API `beginRenderPass()` and `endPass()`.